### ISSUE TITLE
Is it possible to split transformer between CPU and GPU?

### ISSUE THREAD

**Issue by nicolabertoldi** [pytorch/fairseq#802]
Let assume that I have a GPU with a relatively small amount of GB, say 8Gb.
and let assume that my architecture requires 10Gb.

Is it possible to send a portion of the architecture, say the embedding layers, on CPU, while keeping the rest on CPU?

More in general, is it possible to specify that only few parts of a given architecture run on CPU and the rest on GPU?
If yes, how?

**Comment by lematt1991** [comment]
Currently there is not.  Is the model itself not able to fit in GPU memory, or is it model + training samples that won't fit in memory?  If it's the latter, you can try reducing the batch size, and accumulate gradients by increasing `--update-freq`.  This will be semantically equivalent, but will use less GPU memory.

**Comment by nicolabertoldi** [comment]
@lematt1991 

The model itself does not fit in GPU memory.
In my model there are some parts which I do not want to train.
And I would like to keep these parts in CPU.

I looked at the code, and it seems that a nn.Module can work either in GPU or CPU, in general. Is it correct?
If yes, why should be not possible to have two connected nn.Module,  one Module (say Layer1) on CPU and one module (say Layer2) on GPU, paying attention to transform their input/output from/to GPU/CPU using the functions `cuda()` and `cpu()`.

Obviously, this solution if theoretically possible will generate another type of *transformer* architecture, I could called *mixed-device transformer*

**Comment by lematt1991** [comment]
Yeah definitely you should be able to make these changes.  It's just not currently supported in fairseq, and likely will not be.

**Comment by nicolabertoldi** [comment]
@lematt1991 

is there any reason why you say "likely will not be"?

I would like to keep my development as close as possible to what fairseq offers.

**Comment by lematt1991** [comment]
> is there any reason why you say "likely will not be"?

I'm not sure how widely used this would be, but whether such a PR gets merged is not up to me.  

> I would like to keep my development as close as possible to what fairseq offers.

Keeping these changes in a separate file, should make pulling upstream changes pretty easy.

**Comment by nicolabertoldi** [comment]
@lematt1991 
thank you

I am closing the ticket, for now.
